### PR TITLE
Comparisons for number, string, array, and record

### PR DESCRIPTION
**Pull request by jeremyschlatter**
Hello! This is my first PR to Nickel.

I was writing some Nickel configuration today and found myself wanting string comparisons. I found this open [issue](https://github.com/tweag/nickel/issues/1030), and decided to take a shot at it.

Resolves https://github.com/tweag/nickel/issues/1030

### CODE REVIEW

**Review by yannham — APPROVED** [review]
Thanks a lot for taking this on! It looks pretty good, beside a few secondary comments.

File: core/stdlib/std.ncl [hunk 1]
```diff
@@ -501,6 +501,36 @@
...
+            'Greater
+        "%
+      =
+        let refine
+          | [| 'Lesser, 'Equal, 'Greater |] -> [| 'Lesser, 'Equal, 'Greater |] -> [| 'Lesser, 'Equal, 'Greater |]
+          = match {
+            'Equal => fun b => b,
+            a => fun _ => a,
+          }
+          in
```
[the remark this answers is not in the document]
Comment (jeremyschlatter) [reply core/stdlib/std.ncl]: I agree, your version is better. For some reason it hadn't occurred to me to take two arguments but only `match` on one of them. I like it.

And as a bonus, we can drop the contract annotation as you say. My implementation required it, but yours does not.

(With my implementation I got this without the annotation:)

```
error: incompatible types
    ┌─ /Users/jeremy/src/github.com/tweag/nickel/core/stdlib/std.ncl:528:32
    │
528 │           |> fold_right refine 'Equal
    │                                ^^^^^^ this expression
    │
    = Expected an expression of type `[| 'Equal |] -> [| 'Equal |]`
    = Found an expression of type `[| 'Equal; _erows_a |]`
    = These types are not compatible
```

File: core/stdlib/std.ncl [hunk 2]
```diff
@@ -2761,6 +2810,23 @@
         |> std.array.filter (fun { field, value } => f field value)
         |> from_array,
 
+    on
```
[the remark this answers is not in the document]
Comment (jeremyschlatter) [reply core/stdlib/std.ncl:2911]: Yeah, it's not obvious to me what to call this. For what it's worth, Haskell has a function [`on`](https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-Function.html#v:on) that is very similar:

`(b -> b -> c) -> (a -> b) -> a -> a -> c `

The Nickel function here is the same as Haskell's `on` except:

- the arguments are flipped
- `a` is specialized to `{ _ : b }`
- the `(a -> b)` argument is restricted to a call to `std.record.get`

Curious to hear how the bikeshedding goes.

**Comment by jeremyschlatter** [comment]
Thanks @yannham! I made the changes you suggested.

File: core/stdlib/std.ncl (hunk 1 above)
[the remark this answers is not in the document]
Comment (yannham) [reply core/stdlib/std.ncl]: Hmm, the original error is a bit strange. I'll see if I can get a minimal repro; even if there's a good reason for the original bit to not be well-typed (at least with respect to the current type system, as it does look fundamentally fine), the message is puzzling.

**Comment by yannham** [comment]
With 4th of July and unrelated stuff, the weekly was rather empty, so let's bikeshed a bit here. I think `on` works better in Haskell because of the infix notation `` compare `on` length ``. While `on` wouldn't be horrible, I have the following suggestions:

- `apply_on`/`map_on`/`get_map`. I think those are more telling that we map a function over a specific field. However, it's not clear that the function has arity 2 (`get_map` sounds more like `forall a b. String -> (a -> b) -> {_ : a} -> b`). We can add `2` as a suffix but it's less nice
- `zip_on`: there is the precedent of `zip_with`, so `zip` is already used for applying a binary function with arguments coming from two different sources, which is nice. Maybe it should be theoretically `zip_on_with` but it's slightly uglier IMHO.

**Comment by jeremyschlatter** [comment]
I've been thinking about these names for an embarrassingly long amount of time and I'm starting to get semantic satiation 😅

Variants of `zip` and `map` both suggest to me that we'll be getting back a `{_ : b}`, not a `b`.

I considered `at` and  variants of `*_at` but it again makes me kind of expect we'll get back a `{_ : b}` instead of a `b`.

`apply` doesn't have that connotation to me, so it feels better. I agree the 2-arity isn't obvious, but I'm inclined to just accept that. Besides `zip` the only idea I have to communicate it is appending a `2` at the end of the name, but I find that pretty ugly.

I kind of like `apply_to`.

That leaves my top choices as `on` / `apply_on` / `apply_to`.

The use case that motivated me to open this PR in the first place was sorting an array of records by name. Here's how that looks with these options:

```nickel
 array.sort (record.on "name" string.compare)
 array.sort (record.apply_on "name" string.compare)
 array.sort (record.apply_to "name" string.compare)
```

I'm not committed to those, though, and I'm happy for you to unilaterally make a call. You have a lot more invested in nickel than I do.

**Comment by yannham** [comment]
Thanks for your input. I think I like `apply_on`, which combines the `on` and makes it clearer that we select one field (while `to` is a bit too generic IMHO and could relate to the records as well), but is more telling than just `on`. @jeremyschlatter would that be ok to do the renaming directly in this PR? After that we can merge.

**Comment by jeremyschlatter** [comment]
@yannham Done.

**Comment by yannham** [comment]
Great. Thanks again for contributing!

**Pull request merged**